### ISSUE TITLE
incorrect schema generated when using discriminated value and field constraints

### ISSUE THREAD

**Issue by danielkatzan** [pydantic/pydantic#4370]
### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://pydantic-docs.helpmanual.io) and still think this is a bug
- [X] I am confident that the issue is with pydantic  (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))

### Description

When creating a model with an Annotated field which has a discrimination, and combining it with a Field with restrictions, the generated schema doesn't properly takes into account the discriminated values

root cause seems to be similar to this [issue](https://github.com/pydantic/pydantic/issues/4333)

In case there is a Field with any constraints, other annotations are ignored, in this case the discriminated value is ignored due to overriding of the annotation [here](https://github.com/pydantic/pydantic/blob/master/pydantic/schema.py#L1004)

### Example Code

```Python
class Cat(BaseModel):
    pet_type: Literal['cat']

class Dog(BaseModel):
    pet_type: Literal['dog']

Pet = Annotated[Union[Cat, Dog], Field(discriminator='pet_type')]

class PetsWithDiscrimination(BaseModel):
    pets: List[Pet] = Field(title="Have discriminated values")

class PetsWithoutDiscrimination(BaseModel):
    pets: List[Pet] = Field(title="Don't have discriminated values", max_items=2)

PetsWithDiscrimination.schema()  # Contains mapping as expected
PetsWithoutDiscrimination.schema()  # Don't contain mapping

{'title': 'PetsWithDiscrimination',
 'type': 'object',
 'properties': {'pets': {'title': 'Have discriminated values',
   'type': 'array',
   'items': {'discriminator': {'propertyName': 'pet_type',
     'mapping': {'cat': '#/definitions/Cat', 'dog': '#/definitions/Dog'}},
    'anyOf': [{'$ref': '#/definitions/Cat'}, {'$ref': '#/definitions/Dog'}]}}},
 'required': ['pets'],
 'definitions': {'Cat': {'title': 'Cat',
   'type': 'object',
   'properties': {'pet_type': {'title': 'Pet Type',
     'enum': ['cat'],
     'type': 'string'}},
   'required': ['pet_type']},
  'Dog': {'title': 'Dog',
   'type': 'object',
   'properties': {'pet_type': {'title': 'Pet Type',
     'enum': ['dog'],
     'type': 'string'}},
   'required': ['pet_type']}}}


{'title': 'PetsWithoutDiscrimination',
 'type': 'object',
 'properties': {'pets': {'title': "Don't have discriminated values",
   'maxItems': 2,
   'type': 'array',
   'items': {'anyOf': [{'$ref': '#/definitions/Cat'},
     {'$ref': '#/definitions/Dog'}]}}},
 'required': ['pets'],
 'definitions': {'Cat': {'title': 'Cat',
   'type': 'object',
   'properties': {'pet_type': {'title': 'Pet Type',
     'enum': ['cat'],
     'type': 'string'}},
   'required': ['pet_type']},
  'Dog': {'title': 'Dog',
   'type': 'object',
   'properties': {'pet_type': {'title': 'Pet Type',
     'enum': ['dog'],
     'type': 'string'}},
   'required': ['pet_type']}}}
```

### Python, Pydantic & OS Version

```Text
pydantic version: 1.9.0
            pydantic compiled: False
                 install path: /Users/danielkatzan/personal/pydantic/pydantic
               python version: 3.9.13 (main, May 24 2022, 21:13:51)  [Clang 13.1.6 (clang-1316.0.21.2)]
                     platform: macOS-12.1-arm64-arm-64bit
     optional deps. installed: ['devtools', 'dotenv', 'email-validator', 'typing-extensions']
```

### Affected Components

- [ ] [Compatibility between releases](https://pydantic-docs.helpmanual.io/changelog/)
- [ ] [Data validation/parsing](https://pydantic-docs.helpmanual.io/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://pydantic-docs.helpmanual.io/usage/exporting_models/) - `.dict()` and `.json()`
- [X] [JSON Schema](https://pydantic-docs.helpmanual.io/usage/schema/)
- [ ] [Dataclasses](https://pydantic-docs.helpmanual.io/usage/dataclasses/)
- [ ] [Model Config](https://pydantic-docs.helpmanual.io/usage/model_config/)
- [ ] [Field Types](https://pydantic-docs.helpmanual.io/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://pydantic-docs.helpmanual.io/usage/validation_decorator/)
- [ ] [Generic Models](https://pydantic-docs.helpmanual.io/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://pydantic-docs.helpmanual.io/usage/models/) - `construct()`, pickling, private attributes, ORM mode
- [ ] [Settings Management](https://pydantic-docs.helpmanual.io/usage/settings/)
- [ ] [Plugins](https://pydantic-docs.helpmanual.io/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by samuelcolvin** [comment]
Thanks for reporting, this should get fixed in V2 where all validation will be migrated to pydantic-core.

Sadly I don't think it will be possible to fix this before then in V1.10.

**Comment by Kludex** [comment]
Since we did a complete refactor on how we generate the JSON schema, I don't think we'll be fixing it on V1. That said, this is already fixed on V2. 🙏
